src/metadataGeneration/metadataGenerator.ts

```diff
@@ -1,23 +1,52 @@
 import * as mm from 'minimatch';
 import * as ts from 'typescript';
+import { importClassesFromDirectories } from '../utils/importClassesFromDirectories';
 import { ControllerGenerator } from './controllerGenerator';
+import { GenerateMetadataError } from './exceptions';
 import { Tsoa } from './tsoa';
 
 export class MetadataGenerator {
   public readonly nodes = new Array<ts.Node>();
-  public readonly typeChecker: ts.TypeChecker;
-  private readonly program: ts.Program;
+  public typeChecker: ts.TypeChecker;
+  private program: ts.Program;
   private referenceTypeMap: Tsoa.ReferenceTypeMap = {};
   private circularDependencyResolvers = new Array<(referenceTypes: Tsoa.ReferenceTypeMap) => void>();
 
   public IsExportedNode(node: ts.Node) { return true; }
 
-  constructor(entryFile: string, compilerOptions?: ts.CompilerOptions, private readonly ignorePaths?: string[]) {
+  constructor(entryFile: string, private readonly compilerOptions?: ts.CompilerOptions, private readonly ignorePaths?: string[], private readonly controllers?: string[]) {
     this.program = ts.createProgram([entryFile], compilerOptions || {});
     this.typeChecker = this.program.getTypeChecker();
   }
 
   public Generate(): Tsoa.Metadata {
+    if (this.controllers) {
+      this.setProgramToDynamicControllersFiles(this.controllers);
```

Comment:
It would be awesome if we could leave `typeChecker` and `program` as readonly. I think if you move this check into the controller, then you don’t have make those variables mutable.

Reply:
do you mean check this in the c'tor? still need to run all the logic of importClassesFromDirectories to get the classes, and make length check and then to init the program and the type checker 
so if ill move it all to the c'tor it wont be readable

Reply:
You can achieve both goals of readability and immutability by changing:
`private setProgramToDynamicControllersFiles(controllers): void`
to be a pure function that can be used in the constructor, like:
`private createProgramToDynamicControllersFiles(controllers): ts.Program {`

So the constructor would look like:
```
this.program = !!controllers ?
    createProgramToDynamicControllersFiles(...) :
    ts.createProgram([entryFile], compilerOptions || {});
```